### ISSUE TITLE
Aggregation emits multiple events for update of a record part of the aggregation

### ISSUE THREAD

**Issue by praveensrao** [confluentinc/ksql#6616]
I have been trying to code a fairly straightforward use case using kafka / ksqldb and I think I have hit a bug.
 
I've updated the example to talk about a more general use case.
What I have is `pagestats` data with the stats for a given page getting updated in the source database (mongoDB in this case). The data contains :

 - pagename
 - hits
 - state

**My requirement is simple. What I need is an aggregated stream of sum(hits) grouped by state across all pages, as a stream to update my target (in this case it is an in-memory cache) continuously.**

I have a mongoDB kafka connector that streams data from mongoDB using changestreams on a collection into a kafka topic.

Create connector:

```
    curl -X PUT http://praveensrao:8083/connectors/pagestats001/config \
         -H "Content-Type: application/json" -d '{
             "connector.class":"com.mongodb.kafka.connect.MongoSourceConnector",
             "publish.full.document.only":"true",
             "tasks.max":"1",
             "change.stream.full.document":"default",
             "collection":"pagestats001",
             "key.converter.schemas.enable":"true",
             "database":"test",
             "name":"pagestats001",
             "connection.uri":"mongodb://***:***@****:27017/test?replicaSet=****authSource=admin&ssl=true",
             "value.converter.schemas.enable":"false",
             "copy.existing":"false",
             "value.converter":"org.apache.kafka.connect.storage.StringConverter",
             "key.converter":"org.apache.kafka.connect.storage.StringConverter",
             "errors.tolerance": "all",
             "max.message.bytes": "16777216"
        }'
```

 
I insert documents into the mongoDB collection `pagestats001`, and we have messages in the kafka topic `test.pagestats001` this way:
 
```
    ksql> print "test.pagestats001" from beginning;
    Key format: JSON or SESSION(KAFKA_STRING) or HOPPING(KAFKA_STRING) or TUMBLING(KAFKA_STRING) or KAFKA_STRING
    Value format: JSON or KAFKA_STRING
    rowtime: 2020/11/16 03:33:29.780 Z, key: {"_id": {"_data": {"$binary": "gl+x8wgAAAADRjxfaWQAPDZkNmQ3YTFhLWU0NGQtNDI2Yy1iMzMzLTAwYjkzNDY0ZDQ0NAAAWhAEyYMCANg0R8K7k9TJ3Qfd3gQ=", "$type": "00"}}}, value: {"_id": "6d6d7a1a-e44d-426c-b333-00b93464d444", "pagename": "page-1", "state": "ON", "hits": 5}
    rowtime: 2020/11/16 03:34:43.297 Z, key: {"_id": {"_data": {"$binary": "gl+x81EAAAAIRjxfaWQAPGVhZGNjMjMxLTNlZmQtNGMzNy05ZGEzLWQ4MGY3ZTgzOGUzNwAAWhAEyYMCANg0R8K7k9TJ3Qfd3gQ=", "$type": "00"}}}, value: {"_id": "eadcc231-3efd-4c37-9da3-d80f7e838e37", "pagename": "page-2", "state": "ON", "hits": 8}
    rowtime: 2020/11/16 03:35:07.565 Z, key: {"_id": {"_data": {"$binary": "gl+x82UAAAAwRjxfaWQAPGJhNTE4YTQ5LWVhODEtNDY1Zi04ZWY3LWI3ZDZkMzk4ZmYyYwAAWhAEyYMCANg0R8K7k9TJ3Qfd3gQ=", "$type": "00"}}}, value: {"_id": "ba518a49-ea81-465f-8ef7-b7d6d398ff2c", "pagename": "page-1", "state": "SK", "hits": 2}
    rowtime: 2020/11/16 03:35:13.709 Z, key: {"_id": {"_data": {"$binary": "gl+x824AAAABRjxfaWQAPDgwOWZkZDNlLTkxMmQtNDcwNC05NmE0LTIyMWE5OWI0ZTkzZQAAWhAEyYMCANg0R8K7k9TJ3Qfd3gQ=", "$type": "00"}}}, value: {"_id": "809fdd3e-912d-4704-96a4-221a99b4e93e", "pagename": "page-2", "state": "SK", "hits": 7}
```

 
Since the key is not useable for a ksqldb table, doing this based on some documentation online to repartition by _id from the value instead of using the key:
 
```
    CREATE STREAM test_pagestats001_schema (
        _id VARCHAR,
        pagename VARCHAR,
        state VARCHAR,
        hits BIGINT
      ) WITH (
         KAFKA_TOPIC='test.pagestats001', 
         VALUE_FORMAT='JSON'
      );
     
     Message        
    ----------------
     Stream created 
    ----------------
     
    CREATE STREAM test_pagestats001_stream
      WITH (KAFKA_TOPIC='test.pagestats001.byid') AS
      SELECT 
        _id,
        pagename,
        state,
        hits
      FROM test_pagestats001_schema
      PARTITION BY _id
      EMIT CHANGES;
     
     Message                                                
    --------------------------------------------------------
     Created query with ID CSAS_TEST_PAGESTATS001_STREAM_77 
    --------------------------------------------------------
     
    CREATE TABLE test_pagestats001_byid (
         _id VARCHAR PRIMARY KEY, 
        pagename VARCHAR,
        state VARCHAR,
        hits BIGINT
       ) WITH (
         KAFKA_TOPIC='test.pagestats001.byid',
         VALUE_FORMAT='JSON'
      );
     
     Message       
    ---------------
     Table created 
    ---------------
     
    CREATE TABLE test_pagestats001_bystate AS
    SELECT
        state,
        sum(hits) as hits
    FROM test_pagestats001_byid
    GROUP BY state
    emit changes;
     
     Message                                                 
    ---------------------------------------------------------
     Created query with ID CTAS_TEST_PAGESTATS001_BYSTATE_81 
    ---------------------------------------------------------

```
 
Now, I run the below

`    ksql> select * from test_pagestats001_bystate emit changes;
`

Let's say I insert a record into the DB for a state/province "AB"
In the source topic `test.pagestats001` - we have:
`
    rowtime: 2020/11/16 04:07:48.290 Z, key: {"_id": {"_data": {"$binary": "gl+x+xAAAAABRjxfaWQAPDYxNzg0OGVlLTUxNDctNDczZS04MjU3LTE0NzA5ZDg2YzZjOQAAWhAEyYMCANg0R8K7k9TJ3Qfd3gQ=", "$type": "00"}}}, value: {"_id": "617848ee-5147-473e-8257-14709d86c6c9", "pagename": "page-1", "state": "AB", "hits": 2}
`
 
And in table `test_pagestats001_bystate`, we have:
```

    ksql> select * from test_pagestats001_bystate emit changes;
    +-------------------------+-----------------------+
    |STATE                    |HITS                   |
    +-------------------------+-----------------------+
    |AB                       |2                      |

```
 
 
Now, if I insert one more record into the DB for "page-2" for the same state with hits (1), the following event is emitted:
 
`
    |AB                       |3                      |
`
 
This is fine till now.
 
But now comes the problem.
Let's say I update the same record I just inserted as hits(8). Then I get 2 events that are emitted :
 
```

    |AB                       |2                      |
    |AB                       |10                     |

```
 
instead of 1 single event that represents the aggregated value of (10).
***In other words, I was not expecting the event for the aggregated value of (2).***
 
This seems like a basic use case to me, and I think it is a bug.
```
    ksql> version
    Version: 6.0.0
```

**Comment by rodesai** [comment]
@praveensrao this isn't exactly a bug - it's a bit of the implementation leaking into the resulting semantics. What's going on here is this. Let's say you have the following row in your source table:

(K, AB, 1)  // K is the key, AB is the state from above and 1 is the hits

And your current aggregate is:
(AB, 3)

Now, let's say you update your source table by issuing the update:
(K, AB, 8)

To implement this, ksqlDB first does:
REMOVE (K, AB, 1): this operation removes the old value from the source table, and removes it from the table aggregate (by subtracting 1 from the aggregate (making it 2)
INSERT (K, AB, 8): this operation inserts the new value into the source, which gives you the final aggregating result (K, AB, 10)

To fix this, we'd need to update the underlying stream processor (Kafka Streams) to merge the 2 events together when updating a row in the source. I agree this would be worthwhile, but is not trivial. In the meantime would it be possible to handle this case in your application?

**Comment by praveensrao** [comment]
Thanks for the analysis and triage @rodesai 
I may not be able to handle this in the application since I was planning to use it for a scenario where accuracy was very important.
I might have to bypass ksqldb and use kafka streams directly seems like, unless you were referring to a different approach which I could use.

**Comment by mjsax** [comment]
It for sure not a bug also not anything that "leak", but it's how stream processing works. Note that for an infinite stream input, there is no such thing as a final result for a non-windowed aggregation query. And second, you use the `EMIT CHANGES` clause, and thus request the table's changelog stream as result.

Can we close this ticket? Seems there is some miss-understanding how ksqlDB (or actually how stream processing) works.

@pascalconfluent For an application point of view, you could to lookup queries (so-called pull queries) into the current table state. Ie, `SELECT * FROM test_pagestats001_bystate;` (without the `EMIT CHANGES` clause). For this case, you query the current "snapshot" of the result table and you will get a single row per "aggregation group". Of course, the table will be continuously refined when new input data arrives and thus if you repeat the query, you might get a different result.
